### PR TITLE
Fix testLineAgainstWater to work in all circumstances

### LINKED ISSUE #490: testLineAgainstWater() does always collide with the default water level
**Describe the bug**
If you change the water level, testLineAgainstWater() still collides with the default water level, not with the actual water level you just set.

**To Reproduce**
- `srun setWaterLevel(50)` and go into water
- `crun x, y, z = getElementPosition(localPlayer)`
- `crun testLineAgainstWater(x, y, z, x, y, z-10)` returns false, but you are in water
- `crun testLineAgainstWater(x, y, z, x, y, z-60)` returns true, it collides with the water that was on level 0

**Expected behaviour**
To collide with the new water level not the 0 water level that has been changed.

**MTA Client/Server (please complete the following information):**
1.5.3

**Additional context**
From https://bugs.mtasa.com/view.php?id=9364

### LINKED ISSUE #564: testLineAgainstWater fails in some areas of water
**Describe the bug**
in some places within sa testLineAgainstWater will incorrectly return false

The easiest way to see it is to fly around in the map editor and check if the crosshair turns blue when aiming at the water

**To reproduce**
run this and /water to see the problem in water by the sf docks
```lua
addEventHandler("onClientRender",root,
function()
local line1 = testLineAgainstWater(-1576.6, -9.2, -5, -1576.6, -9.2, 5)
dxDrawLine3D(-1576.6, -9.2, -5, -1576.6, -9.2, 5, tocolor(line1 and 0 or 255,0,line1 and 255 or 0), 10)

    local line2 = testLineAgainstWater(-1568.6, -0.7, -5, -1568.6, -0.7, 5)
    dxDrawLine3D(-1568.6, -0.7, -5, -1568.6, -0.7, 5, tocolor(line2 and 0 or 255,0,line2 and 255 or 0), 10)
end
)

addCommandHandler("water", function() setElementPosition(getLocalPlayer(), -1552, -23, 14) end)
```
**Expected behaviour**
For test line against water function to reliably work.

**Screenshots**

**Version**
v1.5.6-release-14414

**Additional context**

From https://bugs.mtasa.com/view.php?id=5857

### PR DESCRIPTION
**Pull request by forkerer**
This pull request aims to fix #564, it adds new way of detecting line/water collision, i built it as a part of CWater* classes.
New mode also works outside of GTA SA original map boundaries, while old one returns false every time there.

Easiest way to test this is to check water detection here:

image

or outside of map boundaries.

New function looks like this:
```lua
bool float float float testLineAgainstWater ( float startX, float startY, float startZ, float endX, float endY, float endZ )
```

Test resource:
[water.zip](https://github.com/multitheftauto/mtasa-blue/files/2408806/water.zip)

Also fixes #490, see https://github.com/multitheftauto/mtasa-blue/pull/568#issuecomment-423830046

EDIT:
*** Legacy mode removed

### CODE REVIEW

**Comment by forkerer** [comment]
Also, this fixes https://github.com/multitheftauto/mtasa-blue/issues/490 too.

image

**Comment by Einheit-101** [comment]
#490 and this one is exactly one and the same bug report and this fix comes in handy for my server

**Comment by qaisjp** [comment]
Looks good. Is there any use in keeping legacy mode at all?

**Comment by AlexTMjugador** [comment]
Backward compatibility could be seen as one use. It might happen that some obscure (and compiled) script relies on the behavior documented in #490 for its purposes, but I find that possibility so improbable. It's not like that has any useful purpose I can think of.

**Comment by forkerer** [comment]
> Looks good. Is there any use in keeping legacy mode at all?

Backward compatibility was my main reasoning, but i gotta agree with AlexTMjugador that it seems unlikely that it would affect any of the currently existing scripts.

**Comment by qaisjp** [comment]
I'm happy to approve this once it's made default (with no legacy mode)!

**Comment by qaisjp** [comment]
I'm happy to approve this once it's made default (with no legacy mode)!

**Comment by forkerer** [comment]
@qaisjp removed legacy mode, also fixed some newly found issues. Tested it pretty extensively now, and can't find any cases where it's not working as intended.

File: Client/game_sa/CWaterManagerSA.cpp [hunk 1]
```diff
@@ -738,9 +803,89 @@ void CWaterManagerSA::SetWaveLevel(float fWaveLevel)
...
+            (vecStart.fY > 3000.0f && vecEnd.fY > 3000.0f))
+        {
+            return false;
+        }
+    }
+
+    std::vector<CWaterZoneSA*> vecZones;
+    GetZonesIntersecting(vecStart, vecEnd, vecZones);
+
+    if (vecZones.size() < 1)
```
Comment (qaisjp) [inline Client/game_sa/CWaterManagerSA.cpp:840]:
```suggestion
    if (vecZones.empty())
```

**Review by qaisjp — APPROVED** [review]
Tiny suggestion, otherwise looks good.

**Pull request merged**